src/main/java/org/broadinstitute/hellbender/engine/MultiVariantDataSource.java

```diff
@@ -0,0 +1,408 @@
+ *
+ * -Targeted queries by one interval at a time. This also requires the files to have been indexed using
+ *  the bundled tool IndexFeatureFile. Targeted queries by one interval at a time are unaffected by
+ *  any intervals for full traversal set via {@link #setIntervalsForTraversal(List)}.
+ */
+public final class MultiVariantDataSource implements GATKDataSource<VariantContext>, AutoCloseable {
+    private static final Logger logger = LogManager.getLogger(MultiVariantDataSource.class);
+
+    /**
+     * List of FeatureDataSource objects aggregated by this MultiVariantDataSource
+     */
+    private List<FeatureDataSource<VariantContext>> featureDataSources = new ArrayList<>();
+
+    /**
+     * Merged VCF header used for this (aggregate) source, derived from the individual soruces.
+     */
+    private VCFHeader mergedHeader;
+
+    /**
+     * SAMSequenceDictionary used for this (aggregate) source, derived from the individual sources.
+     */
+    private SAMSequenceDictionary mergedDictionary;
+
+    /**
+     * Iterator representing an open traversal over this data source initiated via a call to {@link #iterator}
+     * (null if there is no open traversal). We need this to ensure that each iterator is properly closed,
+     * and to enforce the constraint (required by Tribble) that we never have more than one iterator open
+     * over our feature reader.
+     */
+    private MergingIterator<VariantContext> currentIterator;
+
+    /**
+     * Default value for queryLookaheadBases, if none is specified. This is designed to be large enough
+     * so that in typical usage (ie., query intervals with gradually increasing start locations) there will
+     * be a substantial number of cache hits between cache misses, reducing the number of times we need to
+     * repopulate the cache from disk.
+     */
+    public static final int DEFAULT_QUERY_LOOKAHEAD_BASES = 1000;
+
+    /**
+     * Define the set of incompatibilities that are allowed for SequenceDictionaries in Variant input data sources
+     */
+    private final static Set<SequenceDictionaryUtils.SequenceDictionaryCompatibility> allowedSequenceDictionaryIncompatibilities =
+            new HashSet<>(Arrays.asList(
+                    SequenceDictionaryUtils.SequenceDictionaryCompatibility.COMMON_SUBSET,
+                    SequenceDictionaryUtils.SequenceDictionaryCompatibility.SUPERSET,
+                    SequenceDictionaryUtils.SequenceDictionaryCompatibility.NO_COMMON_CONTIGS));
+
+    /**
+     * Creates a FeatureDataSource backed by the provided File and assigns this data source the specified logical
+     * name. We will look ahead the default number of bases ({@link #DEFAULT_QUERY_LOOKAHEAD_BASES}) during queries
+     * that produce cache misses.
+     *
+     * @param featureFile file containing Features
+     * @param name logical name for this data source (may be null)
+     */
+    public void addFeatureDataSource(final File featureFile, final String name) {
```

Comment:
Would requiring that all `FeatureDataSources` be provided at construction time simplify the code at all? If yes, it might be a good change.